nullaway/src/main/java/com/uber/nullaway/handlers/AbstractFieldContractHandler.java

```diff
@@ -0,0 +1,247 @@
+   */
+  protected boolean validateAnnotationSyntax(
+      Set<String> content,
+      NullAway analysis,
+      MethodTree tree,
+      VisitorState state,
+      Symbol.MethodSymbol methodSymbol) {
+    if (content == null) {
+      return true;
+    } else {
+      if (content.isEmpty()) {
+        // we should not allow useless annotations.
+        reportMatch(
+            analysis,
+            state,
+            tree,
+            "empty @"
+                + annotName
+                + " is the default precondition for every method, please remove it.");
+        return false;
+      } else {
+        for (String fieldName : content) {
+          if (fieldName.contains(".")) {
+            if (!fieldName.startsWith(THIS_NOTATION)) {
+              reportMatch(
+                  analysis,
+                  state,
+                  tree,
+                  "currently @"
+                      + annotName
+                      + " supports only class fields of the method receiver: "
+                      + fieldName
+                      + " is not supported");
+              return false;
+            } else {
+              fieldName = trimReceiver(fieldName);
+            }
+          }
+          Symbol.ClassSymbol classSymbol = ASTHelpers.enclosingClass(methodSymbol);
+          Element field = getFieldFromClass(classSymbol, fieldName);
+          if (field == null) {
+            reportMatch(
+                analysis,
+                state,
+                tree,
+                "cannot find field [" + fieldName + "] in class: " + classSymbol.getSimpleName());
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns the field name excluding its receiver (e.g. "this.a" will be "a")
+   *
+   * @param fieldName A class symbol.
+   * @return The class field name.
+   */
+  protected static String trimReceiver(String fieldName) {
+    return fieldName.substring(fieldName.lastIndexOf(".") + 1);
+  }
+
+  /**
+   * Finds a specific field of a class
+   *
+   * @param classSymbol A class symbol.
+   * @param name Name of the field.
+   * @return The class field with the given name.
+   */
+  protected static Element getFieldFromClass(Symbol.ClassSymbol classSymbol, String name) {
+    Preconditions.checkNotNull(classSymbol);
+    for (Element member : classSymbol.getEnclosedElements()) {
+      if (member.getKind().isField()) {
+        if (member.getSimpleName().toString().equals(name)) {
+          return member;
+        }
+      }
+    }
+    Symbol.ClassSymbol superClass = (Symbol.ClassSymbol) classSymbol.getSuperclass().tsym;
+    if (superClass != null) {
+      return getFieldFromClass(superClass, name);
+    }
+    return null;
+  }
+
+  protected static List<Element> getReceiverTreeElements(Tree receiver) {
+    List<Element> elements = new ArrayList<>();
+    if (receiver != null) {
+      elements.add(0, ASTHelpers.getSymbol(receiver));
```

Comment:
I think this is quadratic in the depth of the tree, since each time you insert at index `0` of the `ArrayList` all the existing elements need to be shifted over.  I think you'd be better off constructing the list in the opposite order, and then reversing it at the end.